fsspec/core.py

```diff
@@ -332,48 +332,48 @@ def _un_chain(path, kwargs):
     # [[url, protocol, kwargs], ...]
     out = []
     previous_bit = None
-    previous_protocol = None
     for bit in reversed(bits):
         protocol = split_protocol(bit)[0] or "file"
         cls = get_filesystem_class(protocol)
         extra_kwargs = cls._get_kwargs_from_urls(bit)
-        kws = kwargs.get(split_protocol(bit)[0] or "file", {})
+        kws = kwargs.get(protocol, {})
         kw = dict(**extra_kwargs, **kws)
+        bit = cls._strip_protocol(bit)
         if (
             protocol in {"blockcache", "filecache", "simplecache"}
             and "target_protocol" not in kw
         ):
-            bit = previous_bit.replace(previous_protocol, protocol, 1)
+            bit = previous_bit
         out.append((bit, protocol, kw))
         previous_bit = bit
-        previous_protocol = protocol
     out = list(reversed(out))
-    # We should only do the url rewrite if the cache is in the middle of the chain
```

Comment:
In get_fs_token_paths it is easier to expand just the first fs paths. For that it is convenient to bring the target fs path to the cache fs.
Since this is an internal function only used by url_to_fs and get_fs_token_paths, no negative impact is expected.